// @arwes/animator/build/cjs/internal/createAnimatorMachine/createAnimatorMachine.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAnimatorMachine = void 0;
const tools_1 = require("@arwes/tools");
const constants_1 = require("../../constants");
const index_1 = require("../../internal/createAnimatorManager/index");
const createAnimatorMachine = (node, initialState) => {
    let state = initialState;
    const statesMap = {
        [constants_1.ANIMATOR_STATES.exited]: {
            onActions: {
                [constants_1.ANIMATOR_ACTIONS.enter]: constants_1.ANIMATOR_STATES.entering,
                [constants_1.ANIMATOR_ACTIONS.setup]: () => {
                    const settings = node.control.getSettings();
                    if (node.parent) {
                        const parentSettings = node.parent.control.getSettings();
                        switch (node.parent.state) {
                            case constants_1.ANIMATOR_STATES.entering: {
                                if (parentSettings.combine || settings.merge) {
                                    node.parent.manager.enterChildren([node]);
                                }
                                break;
                            }
                            // If the parent has already entered, enter the incoming children whether
                            // they have "merge" setting or the parent is in "combine" setting.
                            case constants_1.ANIMATOR_STATES.entered: {
                                node.parent.manager.enterChildren([node]);
                                break;
                            }
                        }
                    }
                    else {
                        const isActive = settings.active === undefined || settings.active;
                        if (isActive) {
                            return constants_1.ANIMATOR_STATES.entering;
                        }
                    }
                }
            }
        },
        [constants_1.ANIMATOR_STATES.entering]: {
            onEntry: {
                execute: () => {
                    const { combine } = node.control.getSettings();
                    const children = combine
                        ? Array.from(node.children)
                        : Array.from(node.children).filter(child => child.control.getSettings().merge);
                    node.manager.enterChildren(children);
                },
                schedule: () => {
                    const { duration } = node.control.getSettings();
                    return {
                        duration: (duration.delay + duration.enter) || 0,
                        action: constants_1.ANIMATOR_ACTIONS.enterEnd
                    };
                }
            },
            onActions: {
                [constants_1.ANIMATOR_ACTIONS.enterEnd]: constants_1.ANIMATOR_STATES.entered,
                [constants_1.ANIMATOR_ACTIONS.exit]: constants_1.ANIMATOR_STATES.exiting,
                [constants_1.ANIMATOR_ACTIONS.refresh]: () => {
                    const settings = node.control.getSettings();
                    const childrenExited = Array
                        .from(node.children)
                        .filter(child => child.state === constants_1.ANIMATOR_STATES.exited);
                    if (settings.combine) {
                        node.manager.enterChildren(childrenExited);
                    }
                    else {
                        const childrenMerged = childrenExited
                            .filter(child => child.control.getSettings().merge);
                        node.manager.enterChildren(childrenMerged);
                    }
                }
            }
        },
        [constants_1.ANIMATOR_STATES.entered]: {
            onEntry: {
                execute: () => {
                    const { combine } = node.control.getSettings();
                    if (combine) {
                        return;
                    }
                    const children = Array
                        .from(node.children)
                        .filter(child => !child.control.getSettings().merge);
                    node.manager.enterChildren(children);
                }
            },
            onActions: {
                [constants_1.ANIMATOR_ACTIONS.exit]: constants_1.ANIMATOR_STATES.exiting,
                [constants_1.ANIMATOR_ACTIONS.refresh]: () => {
                    const childrenExited = Array
                        .from(node.children)
                        .filter(child => child.state === constants_1.ANIMATOR_STATES.exited);
                    node.manager.enterChildren(childrenExited);
                }
            }
        },
        [constants_1.ANIMATOR_STATES.exiting]: {
            onEntry: {
                execute: () => {
                    Array.from(node.children).forEach(child => {
                        if (child.state === constants_1.ANIMATOR_STATES.entering || child.state === constants_1.ANIMATOR_STATES.entered) {
                            child.send(constants_1.ANIMATOR_ACTIONS.exit);
                        }
                        else if (child.state === constants_1.ANIMATOR_STATES.exited) {
                            child.scheduler.stopAll();
                        }
                        // If the child is EXITING, it will go to EXITED soon.
                    });
                },
                schedule: () => ({
                    duration: node.control.getSettings().duration.exit || 0,
                    action: constants_1.ANIMATOR_ACTIONS.exitEnd
                })
            },
            onActions: {
                [constants_1.ANIMATOR_ACTIONS.exitEnd]: constants_1.ANIMATOR_STATES.exited,
                [constants_1.ANIMATOR_ACTIONS.enter]: constants_1.ANIMATOR_STATES.entering
            }
        },
        '*': {
            onActions: {
                [constants_1.ANIMATOR_ACTIONS.update]: () => {
                    var _a, _b;
                    const settings = node.control.getSettings();
                    if (settings.manager !== node.manager.name) {
                        (_b = (_a = node.manager).destroy) === null || _b === void 0 ? void 0 : _b.call(_a);
                        node.manager = (0, index_1.createAnimatorManager)(node, settings.manager);
                    }
                    if (!node.parent) {
                        const isActive = settings.active === true ||
                            settings.active === undefined;
                        if ((state === constants_1.ANIMATOR_STATES.exited || state === constants_1.ANIMATOR_STATES.exiting) && isActive) {
                            return constants_1.ANIMATOR_STATES.entering;
                        }
                        else if ((state === constants_1.ANIMATOR_STATES.entered || state === constants_1.ANIMATOR_STATES.entering) && !isActive) {
                            return constants_1.ANIMATOR_STATES.exiting;
                        }
                    }
                }
            }
        }
    };
    const transition = (newState) => {
        if (!newState || state === newState) {
            return;
        }
        state = newState;
        const { onEntry } = statesMap[state] || {};
        const { onTransition } = node.control.getSettings();
        node.scheduler.stopAll();
        if (onEntry === null || onEntry === void 0 ? void 0 : onEntry.execute) {
            onEntry.execute();
        }
        if (onEntry === null || onEntry === void 0 ? void 0 : onEntry.schedule) {
            const task = onEntry.schedule();
            node.scheduler.start(task.duration, () => send(task.action));
        }
        onTransition === null || onTransition === void 0 ? void 0 : onTransition(node);
        for (const subscriber of node.subscribers) {
            subscriber(node);
        }
    };
    const processAction = (procedure) => {
        if (procedure === undefined) {
            return;
        }
        if (typeof procedure === 'string') {
            transition(procedure);
        }
        else {
            const newState = procedure();
            if (newState) {
                transition(newState);
            }
        }
    };
    const getState = () => state;
    const send = (action) => {
        var _a, _b, _c, _d;
        // In non-browser environments, there are no transitions.
        if (!tools_1.IS_BROWSER) {
            return;
        }
        processAction((_b = (_a = statesMap[state]) === null || _a === void 0 ? void 0 : _a.onActions) === null || _b === void 0 ? void 0 : _b[action]);
        processAction((_d = (_c = statesMap['*']) === null || _c === void 0 ? void 0 : _c.onActions) === null || _d === void 0 ? void 0 : _d[action]);
    };
    const machine = Object.freeze({ getState, send });
    return machine;
};
exports.createAnimatorMachine = createAnimatorMachine;
